src/hypothesis/strategies.py

```diff
@@ -982,32 +987,61 @@ def fractions(min_value=None, max_value=None, max_denominator=None):
 
     If max_denominator is not None then the absolute value of the denominator
     of any generated values is no greater than max_denominator. Note that
-    max_denominator must be at least 1.
+    max_denominator must be None or a positive integer, and the denominator
+    of min_value and max_value may not be greater than max_denominator.
 
     """
+    if min_value is not None:
+        min_value = Fraction(min_value)
+    if max_value is not None:
+        max_value = Fraction(max_value)
+
     check_valid_bound(min_value, 'min_value')
     check_valid_bound(max_value, 'max_value')
     check_valid_interval(min_value, max_value, 'min_value', 'max_value')
 
-    check_valid_integer(max_denominator)
-    if max_denominator is not None and max_denominator < 1:
-        raise InvalidArgument(
-            u'Invalid denominator bound %s' % max_denominator
-        )
-
-    denominator_strategy = integers(min_value=1, max_value=max_denominator)
-
     def dm_func(denom):
-        max_num = max_value * denom if max_value is not None else None
-        min_num = min_value * denom if min_value is not None else None
+        """Take denom, construct numerator strategy, and build fraction."""
+        # Four cases of algebra to get integer bounds and scale factor.
+        min_num, max_num = None, None
+        if max_value is None and min_value is None:
+            pass
+        elif min_value is None:
+            max_num = denom * max_value.numerator
+            denom *= max_value.denominator
+        elif max_value is None:
+            min_num = denom * min_value.numerator
+            denom *= min_value.denominator
+        else:
+            min_num = min_value.numerator * denom * max_value.denominator
+            max_num = max_value.numerator * denom * min_value.denominator
+            denom *= min_value.denominator * max_value.denominator
 
         return builds(
             Fraction,
             integers(min_value=min_num, max_value=max_num),
             just(denom)
         )
 
-    return denominator_strategy.flatmap(dm_func)
+    if max_denominator is None:
+        if min_value is not None and min_value == max_value:
+            return just(min_value)
+        return integers(min_value=1).flatmap(dm_func)
+
+    check_valid_integer(max_denominator)
+    if max_denominator < 1:
+        raise InvalidArgument('Denominator=%r must be >= 1' % max_denominator)
+    if min_value is not None and min_value.denominator > max_denominator:
+        raise InvalidArgument('min_value=%r incompatible with max_denominator'
+                              '=%r' % (min_value, max_denominator))
+    if max_value is not None and max_value.denominator > max_denominator:
+        raise InvalidArgument('max_value=%r incompatible with max_denominator'
+                              '=%r' % (max_value, max_denominator))
```

Comment:
It's true that this would not previous raise an exception, but nor has it ever worked correctly - if you tried, the denominator bound would be respected but the value bounds would not!

There are some cases where with a value-bound-denominator greater than the max_denominator, no examples are possible: eg `fractions('1/3', '2/3', 1)` has no valid output and nor has `fractions(.5, .5, 1)`.

I see three choices:

1. Allow some examples to be outside the specified bounds, or allow bounds with no possible examples.
2. Validate the value-bounds with respect to the precision bound.  (I've done this because I like it - the way the various intervals all fall into place is lovely).  Note that value bounds may be unrepresentable with fractions of max_denominator; it all still works perfectly.
3. Try to validate whether there are any possible outputs in cases where a bound has too high a denominator.  If `max_value - min_value >= Fraction(1, max_denominator)`, this is trivially true.  If not, I haven't yet been able to get possible example denominator values in less than linear time of the greatest denominator (by checking each).  The trivial case of equal value bounds with an illegal denominator is obvious but not much use.

The whole point of this pull is that I don't like option one, and three is similar but less consistent!